site/src/pages/RoadmapPage/sidebar/MajorRequiredCourseList.tsx

```diff
@@ -182,46 +185,84 @@ const MajorRequiredCourseList: FC = () => {
+        isMulti
         options={majorSelectOptions}
-        // must be NULL, not undefined, to clear the field for unselected values
-        value={majorSelectOptions.find((o) => o.value === selectedMajor) ?? null}
+        value={selectedMajors.map((m) => majorSelectOptions.find((o) => o.value.id === m.major.id)!)}
         isDisabled={majorsLoading}
         isLoading={majorsLoading}
-        onChange={(data) => {
-          if (data!.value.id === selectedMajor?.id) return;
-          dispatch(setRequirements([])); // set to empty immediately because otherwise it's out of date
-          dispatch(setMajor(data!.value));
-          dispatch(setSpecialization(null));
-          saveMajor(data!.value);
+        onChange={(selections) => {
+          const newMajors = selections?.map((s) => s.value) || [];
+          const currentMajorIds = selectedMajors.map((m) => m.major.id);
+          currentMajorIds.forEach((id) => {
+            if (!newMajors.find((m) => m.id === id)) {
+              dispatch(removeMajor(id));
+            }
+          });
+          newMajors.forEach((major) => {
+            if (!currentMajorIds.includes(major.id)) {
+              dispatch(addMajor(major));
+              fetchRequirements(major.id);
+            }
+          });
+          const updatedMajors = newMajors.map((major) => ({
+            major,
+            specialization: selectedMajors.find((m) => m.major.id === major.id)?.specialization || null,
+            requirements: selectedMajors.find((m) => m.major.id === major.id)?.requirements || [],
+          }));
+          saveMajors(updatedMajors);
         }}
         className="ppc-combobox"
         classNamePrefix="ppc-combobox"
-        placeholder="Select a major..."
+        placeholder="Select majors..."
         theme={(t) => comboboxTheme(t, isDark)}
       />
-      {selectedMajor && !!selectedMajor.specializations.length && (
-        <Select
-          options={specSelectOptions}
-          value={specSelectOptions.find((o) => o.value === selectedSpec)}
-          key={selectedMajor.id} // force re-render on changing major
-          isDisabled={specsLoading}
-          isLoading={specsLoading}
-          onChange={(data) => {
-            if (data!.value.id === selectedSpec?.id) return;
-            setResultsLoading(true);
-            dispatch(setRequirements([])); // set to empty immediately because otherwise it's out of date
-            dispatch(setSpecialization(data!.value));
-            saveSpecialization(data!.value);
-          }}
-          className="ppc-combobox"
-          classNamePrefix="ppc-combobox"
-          placeholder="Select a specialization..."
-          theme={(t) => comboboxTheme(t, isDark)}
-        />
-      )}
-      {selectedMajor && (!selectedMajor.specializations.length || selectedSpec) && (
-        <>{resultsLoading ? loadingIcon : <ProgramRequirementsList requirements={requirements} />}</>
-      )}
+
+      {selectedMajors.map((majorWithSpec) => (
```

Comment:
This is way too much code to have in one component. If I recall correctly, we discussed that the spec selection and requirements display should be extracted into its own component, and that each of those should handle state.

Having all the code through a map makes this difficult to read. I'd name the separate component something like `<SingleMajorCourseList/>` or something of that manner.